### PR TITLE
Refactor `TxOut` Aeson instances

### PR DESCRIPTION
**Pull request by newhoggy**
# Changelog

```yaml
- description: |
    Refactored TxOut JSON parsing to eliminate ~150 lines of code duplication across era cases
    and added comprehensive test coverage for all TxOut JSON serialization scenarios.
# uncomment types applicable to the change:
  type:
  - refactoring    # QoL changes
  - test           # fixes/modifies tests
  - documentation  # change in code docs, haddocks...
# uncomment at least one main project this PR is associated with
  projects:
  - cardano-api
```

# Context

This PR refactors the `TxOut` JSON parsing implementation to eliminate significant code duplication that existed across different Cardano era cases (Babbage, Conway, Dijkstra). The original implementation had nearly identical parsing logic repeated for each era, making the code harder to maintain and more prone to inconsistencies.

Additionally, comprehensive test coverage was added to ensure the refactored JSON instances work correctly across all eras and contexts, including edge cases and error scenarios.

Resolves https://github.com/IntersectMBO/cardano-api/issues/926

# How to trust this PR

The refactoring preserves exact behavioral compatibility while consolidating duplicated code:

**Key improvements:**
- Extracted ~150 lines of duplicated inline datum parsing logic into reusable helper functions
- Created `parseBabbageOnwardsTxOut` to unify parsing for Babbage+ eras
- Introduced `parseInlineDatum` helper to handle inline datum validation consistently
- Consolidated `reconcileBabbage` and `reconcileConway` into single `reconcileDatums` function
- Added `caseBabbageOnlyOrConwayEraOnwards` for explicit era-based branching with compiler-enforced exhaustiveness

**Test coverage added:**
- 37 new comprehensive JSON tests covering all eras (Shelley through Conway)
- Roundtrip property tests for both `CtxTx` and `CtxUTxO` contexts
- Edge case tests for supplemental datums, null field handling, and field validation
- Error case tests for conflicting datums, mismatched hashes, and invalid data
- Specialized generators for different datum types (none, hash, supplemental, inline)

**Documentation improvements:**
- Added extensive inline documentation explaining design decisions and assumptions
- Documented the critical distinction between Babbage's `scriptDataJsonToHashable` vs Conway+'s `scriptDataFromJson`
- Noted remaining duplication in `CtxUTxO` instance for future refactoring

Run the test suite to verify all functionality is preserved:
```bash
# Run full test suite
cabal test cardano-api-test

# Run TxOut-specific JSON tests
cabal test cardano-api-test --test-options="-p TxOut.Json"
```

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [x] Self-reviewed the diff

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Tx/Internal/Output.hs [hunk 1]
```diff
@@ -455,133 +455,88 @@ instance IsShelleyBasedEra era => FromJSON (TxOut CtxTx era) where
...
+      alonzoTxOut <- alonzoTxOutParser (convert w) o
+      inlineDatum <- parseInlineDatum w o
+      mReferenceScript <- o .:? "referenceScript"
+      reconcileDatums w alonzoTxOut inlineDatum mReferenceScript
+
+    -- | Parse inline datum fields from JSON object
+    -- Handles both inlineDatumhash and inlineDatum fields, validating they match
+    parseInlineDatum
+      :: BabbageEraOnwards era
+      -> Aeson.Object
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/Tx/Internal/Output.hs]: **Type-level constraint provides exhaustiveness checking**

By taking `BabbageEraOnwards era` as a parameter instead of `ShelleyBasedEra era`, this function gains an important benefit: the compiler can verify exhaustiveness.

`BabbageEraOnwards` is a GADT with exactly 3 constructors (Babbage, Conway, Dijkstra). If a new era is added to this GADT in the future, any pattern match on it (like in `eraName` or `parseInlineDatum`) will fail to compile unless explicitly handled.

This prevents bugs where new eras are added but parsing logic is forgotten.
